Add Runner for Integration Tests for the SOM language

rhys-h-walker opened the pull request:
This pull request adds the implementation of the test runner for the integration tests previously added in #129. The runner uses PyTest.


To run the tests, one needs to set a few environment variables.
In the current state, this should be something like:

```bash
export VM=./som.sh
export CLASSPATH=./core-lib/Smalltalk
export AWFY=./core-lib/Examples/AreWeFastYet/Core
TEST_EXPECTATIONS=integration-tests.yml python3 -m pytest
```

The file specified with `TEST_EXPECTATIONS` lists the tests that are expected to fail for a specific SOM implementation. This is needed, because various language features are currently not specified, and yksom made its own choices.

### Core-Lib Updates in the SOM Implementations

| Status 	| SOM        	| PR                                           	|
|:------:	|------------	|----------------------------------------------	|
| :white_check_mark: | SOM (java) 	| https://github.com/SOM-st/som-java/pull/40	|
| :white_check_mark: | TruffleSOM 	| https://github.com/SOM-st/TruffleSOM/pull/235	|
| :white_check_mark: | PySOM      	| https://github.com/SOM-st/PySOM/pull/67 |
| :white_check_mark: | JsSOM      	| https://github.com/SOM-st/JsSOM/pull/21 |
| :white_check_mark: | SOM++      	| https://github.com/SOM-st/SOMpp/pull/76 |
|  | SOM-RS     	| https://github.com/Hirevo/som-rs/pull/ |

rhys-h-walker committed:
Implement test_runner.py

This is a PyTest-based runner for the integration tests of SOM.
Inspired by lang_tester https://github.com/softdevteam/lang_tester

- updated CI to run Black and PyLint, and integration tests

- it can generate yaml file with currently failing tests. It can function as a TEST_EXCEPTIONS file with failing tests and passing tests, added and removed as needed. It should not be blindly followed but makes the generation of a tags file very easy.

- it comes with additional information like number passed/skipped and which ones exactly have been changed.

- Tests are case insensitive unless instructed not to be through `case_sensitive: true` in the tests.

- a test can be expected to fail. This allows for testing of test_runner features. Not ideal but makes sure case_sensitivity works correctly.

smarr commented on IntegrationTests/conftest.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    if GENERATE_REPORT:
+        print("Generating report for the test run")
+        os.makedirs(GENERATE_REPORT_LOCATION, exist_ok=True)
```
Why does this need to be a directory? Could `GENERATE_REPORT` not just take the file path of the file that is to be created?

smarr commented on IntegrationTests/conftest.py:
```diff
@@ -0,0 +1,105 @@
+            "unsupported": unsupported,
+            "do_not_run": do_not_run,
+        }
+        print(f"Report location {GENERATE_REPORT_LOCATION}/report.txt")
```
Inconsistent to what you actually do below. `GENERATE_REPORT` being a file name, it avoids this issue. It also avoids hardcoding any file names at all.

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+
+def debug(message):
+    """
+    Take a string as a mesasage and output if DEBUG is true
```
```suggestion
    Take a string as a message and output if DEBUG is true
```

rhys-h-walker replied:
method removed

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+        print(message)
+
+
+def debug_list(message_list, prefix="", postfix=""):
```
do we really need these debug methods? I would rather avoid them. `DEBUG` itself is something I'd probably avoid.

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+def locate_tests(path, test_files):
+    """
+    Locate all test files that exist in the given directory
+    Ignore any tests which are in the ignoredTests directory
```
`ignoredTests` is not a thing anymore, is it?

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+    """
+    Recursively read all sub directories
+    Path is the directory we are currently in
+    testFiles is the list of test files we are building up
```
```suggestion
    test_files is the list of test files we are building up
```

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+        "name": test_file,
+        "stdout": [],
+        "stderr": [],
+        "customCP": "NaN",
```
This doesn't match the readme:

```suggestion
        "custom_classpath": None,
```

But why is this `NaN`? Couldn't this be something more conventional like `None`?

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+            comment_lines = comment.split("\n")
+            for line in comment_lines:
+                if "custom_classpath" in line:
+                    test_info_dict["customCP"] = line.split("custom_classpath:")[
```
```suggestion
                    test_info_dict["custom_classpath"] = line.split("custom_classpath:")[
```

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+                test_info_dict["name"],
+                test_info_dict["stdout"],
+                test_info_dict["stderr"],
+                test_info_dict["customCP"],
```
```suggestion
                test_info_dict["custom_classpath"],
```

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+if external_vars.TEST_EXCEPTIONS:
+    with open(f"{external_vars.TEST_EXCEPTIONS}", "r", encoding="utf-8") as file:
+        yamlFile = yaml.safe_load(file)
+        external_vars.known_failures = yamlFile["known_failures"]
```
These and the related accesses are going to fail, if the accessed keys are not in the file, right? Would be more robust if this doesn't fail, and instead just defaults to an empty list or so.

rhys-h-walker replied:
Added checks for key inclusion in yaml file now.

https://github.com/rhys-h-walker/SOM/commit/03b8dbf0973ca84e50da64fc93f4c7af6cc1caeb

rhys-h-walker commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+        # Check if the file is in the ignored tests (Check via path, multiple tests named test.som)
+        with open(file_path, "r", encoding="utf-8") as f:
+            contents = f.read()
+            if "VM" in contents:
```
Yes, the comment above does say it should be, must have been missed

rhys-h-walker commented on IntegrationTests/conftest.py:
```diff
@@ -0,0 +1,99 @@
+            "do_not_run": do_not_run,
+        }
+        print(f"Report location {GENERATE_REPORT_LOCATION}")
+        with open(f"{GENERATE_REPORT_LOCATION}", "w", encoding="utf-8") as f:
```
GENERATE_REPORT simplified down to one variable.

rhys-h-walker commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+
+    debug(f"Running test: {name}")
+
+    result = subprocess.run(
```
Just reread what you write, so it passes on other SOMs just not SOM++?

rhys-h-walker replied:
I see now, erm I can have a look into what I could do

smarr replied:
Yeah, the different underlying systems might do slightly different things.
In the worst case, maybe we need to ask Python to treat it as a binary stream, and then try to decode it manually. Not sure what the best option is. But ideally the tester is robust enough to deal with what every broken things a broken SOM generates, without itself causing unexpected exceptions like the `UnicodeDecodeError`.

rhys-h-walker replied:
Hm, I can catch it in a try block and maybe then attempt to decode from there. I'll need to look into how the error is raised within subprocess.run, ideally I wouldn't have to convert all outputs to a Binary and could just single out the failing ones

rhys-h-walker replied:
So it won't generate a result object if the subprocess.run fails. I've wrapped it in a try/catch block and skips the test.

It will still report if the test passes for any reason. For now any test will skip on a Unicode error instead of just failing.

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,365 @@
+        result.stderr = str(result.stderr).lower()
+
+    # Produce potential error messages now and then run assertion
+    error_message = f"""
```
Ah, very nice!

This seems to work nicely.

```
E             Stdout diff    :
E             + this is som++
E               0
E               1
E               4294967295
E               4294967295
E               0
E               4294967295
E             - 0
E             - 4294967295
E             + 256
E             + 255
E               0
E               1
E               -1
E               -1
E               0
E               -1
E             - 0
E             - -1+ 256
E             + 255
```

Though, the per test report is now huge.
One simplification that would be nice is to show only relevant things.
So, if stderr has nothing expected, and the result is the expected, then that category of outputs could probably be omitted.

I am also wondering whether we actually need both, the two expected and the actual output as well as the diff. And, might the diff be enough?
And, if we only have the diff, could that one also be prefixed with line numbers?

Thanks a lot!
Stefan

smarr replied:
> The issue here is that SOMs do not necessarily output to the correct output channel all of the time.

I consider this a SOM bug. So, tests should be failing for these cases.


> Line numbers are not perfect and will increment even when just showing a Diff but I think it's obvious enough what is going on.

hm, yeah, line numbers for a diff are not useful like that, indeed.

rhys-h-walker replied:
Removed line number diffs

rhys-h-walker replied:
Will change the success logic to now just check if exp_stdout is in stdout and exp_stderr is in stderr

rhys-h-walker replied:
Seems at least in SOM++ it was not an issue

rhys-h-walker commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+
+    # Check if each line in stdout and stderr is in the expected output
+    for line in expected_std_out:
+        if not any(line in out_line for out_line in std_out):
```
This now checks the order:

https://github.com/rhys-h-walker/SOM/blob/aa6fa2da17a6490040a774fda6be0238edaaa836/IntegrationTests/test_runner.py#L137C1-L207C17

It's a bit longer because we don't know if SOM errors will actually write to stderr in all cases. But works properly now I think.

rhys-h-walker committed:
SOM UnicodeDecodeError:  now more robust against a failure in decode a string.

- skip test on Unicode Failure

rhys-h-walker commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,352 @@
+            external_vars.tests_failed_unexpectedly.append(name)
+        assert (
+            test_pass_bool
+        ), f"Error on test, {name} expected to pass: {error_message}"
```
Shouldn't be needed, any test that should pass would go through that check and either pass/fail as normal. Anything that we expect to fail will just pass through and not report an error but if they pass for any reason will be picked up by the above if statements.

rhys-h-walker committed:
Further improvements

- Removed Print statements leftover
- Updated error handling to just show diff unless no diff is available
- EXECUTABLE changed to be VM
- Added some basic test_runner tests
- Make a diff reverted to just be a diff and not have line numbers
- be strict on inclusion in stdout/stderr
- Removed the lang_tests and adjusted the test_runner logic to now print a proper error on /Tests folder not found
- Fix conftest to output a string rather than an object for exitstatus
- Fixed error with missing or empty tag file, addded test to assert that the yaml files are working correctly
- Added an additional test to check the robustness of read_test_exceptions
- Fixed an error where specifying case_sensitive: False would not register as False but True instead
- Added a new tests to test the discovery of lang_tester valid tests inside a directory
- Pathname for TEST_EXCEPTIONS now just takes a filename if it is located within IntegrationTests.
- Adjusted test_test_diccovery to use relative path names rather than hard coded
- There are now two options for running pytest, normally and execute just the som tests and with the argument -m tester which will run the test_runner tests located in test_tester
- Added deselect message when tests are deselected due to runner choice
- Filepaths must now be given from IntegrationTests onwards so: Tests/test.som
- Updated to now use relative paths rather than hardcoded for TEST_EXCEPTIONS

smarr commented:
Another small issue:

I have test that passes, but was expected to fail. Reporting this should be minimal, and does not need to show the whole stuff.

Currently, it shows this:

```
E           AssertionError: Test core-lib/IntegrationTests/Tests/integer_asdouble.som is in failing_as_unspecified but passed
E
E             Expected stdout:
E             1|    0.0
E             2|    1.0
E             3|    -1.0
E             4|    1.8446744073709552e19
E             5|    -1.8446744073709552e19
E             Given stdout   :
E             1|    0.0
E             2|    1.0
E             3|    -1.0
E             4|    1.8446744073709552e19
E             5|    -1.8446744073709552e19
E             6|
E             Expected stderr:
E
E             Given stderr   :
E             1|
E             Command used   : ./som.sh -cp Smalltalk core-lib/IntegrationTests/Tests/integer_asdouble.som
E             Case sensitive : False
E             Stdout diff    :
E               0.0
E               1.0
E               -1.0
E               1.8446744073709552e19
E             - -1.8446744073709552e19+ -1.8446744073709552e19
E             ?                       +
E
E             Stderr diff    :
E
E
E           assert False
```

But all that's needed is:
`E           AssertionError: Test core-lib/IntegrationTests/Tests/integer_asdouble.som is in failing_as_unspecified but passed`


Also, it looks like there's something odd with the diff:

`E             - -1.8446744073709552e19+ -1.8446744073709552e19`

smarr commented on IntegrationTests/Tests/case_insensitive.som:
```diff
@@ -0,0 +1,20 @@
+
+"
+
+case_insensitive = (
```
Hm, was this designed to be a test for the runner?
I think this is a very good idea, and we should expand on this.
But, we need to find a way of doing that that doesn't make this SOM-dependent.
And we do not want these tests to be ending up in the set of integration tests for SOMs. So, we want to clearly separate tests for the test_runner, and the SOMs.

smarr replied:
Perhaps, you could just write pytest unit tests for the relevant methods?

rhys-h-walker replied:
Yes I just need to check the inputs and outputs of the functions so shouldn't be too challenging. I think it may go in a different python file though to avoid confusion.

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,370 @@
+    if str(name) in external_vars.do_not_run:
+        pytest.skip("Test included in do_not_run")
+
+    if custom_classpath != "None":
```
Hm, why is this a string "None", instead of just a `None`? 
I think there's a bug else where.

rhys-h-walker commented:
> Another small issue:
> 
> I have test that passes, but was expected to fail. Reporting this should be minimal, and does not need to show the whole stuff.
> 
> Currently, it shows this:
> 
> ```
> E           AssertionError: Test core-lib/IntegrationTests/Tests/integer_asdouble.som is in failing_as_unspecified but passed
> E
> E             Expected stdout:
> E             1|    0.0
> E             2|    1.0
> E             3|    -1.0
> E             4|    1.8446744073709552e19
> E             5|    -1.8446744073709552e19
> E             Given stdout   :
> E             1|    0.0
> E             2|    1.0
> E             3|    -1.0
> E             4|    1.8446744073709552e19
> E             5|    -1.8446744073709552e19
> E             6|
> E             Expected stderr:
> E
> E             Given stderr   :
> E             1|
> E             Command used   : ./som.sh -cp Smalltalk core-lib/IntegrationTests/Tests/integer_asdouble.som
> E             Case sensitive : False
> E             Stdout diff    :
> E               0.0
> E               1.0
> E               -1.0
> E               1.8446744073709552e19
> E             - -1.8446744073709552e19+ -1.8446744073709552e19
> E             ?                       +
> E
> E             Stderr diff    :
> E
> E
> E           assert False
> ```
> 
> But all that's needed is: `E AssertionError: Test core-lib/IntegrationTests/Tests/integer_asdouble.som is in failing_as_unspecified but passed`
> 
> Also, it looks like there's something odd with the diff:
> 
> `E - -1.8446744073709552e19+ -1.8446744073709552e19`

Assertions that fail only by being in unspecified/known/unuspported now just show that they passed and are located inside of the tags file.

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,665 @@
+    """
+    if filename:
+        with open(f"{filename}", "r", encoding="utf-8") as file:
+            yaml_file = yaml.safe_load(file)
+
+            if yaml_file is not None:
+                if "known_failures" in yaml_file.keys():
+                    external_vars.known_failures = yaml_file["known_failures"]
+                    if external_vars.known_failures is None:
+                        external_vars.known_failures = []
+                else:
+                    external_vars.known_failures = []
+
+                if "failing_as_unspecified" in yaml_file.keys():
+                    external_vars.failing_as_unspecified = yaml_file[
+                        "failing_as_unspecified"
+                    ]
+                    if external_vars.failing_as_unspecified is None:
+                        external_vars.failing_as_unspecified = []
+                else:
+                    external_vars.failing_as_unspecified = []
+
+                if "unsupported" in yaml_file.keys():
+                    external_vars.unsupported = yaml_file["unsupported"]
+                    if external_vars.unsupported is None:
+                        external_vars.unsupported = []
+                else:
+                    external_vars.unsupported = []
+
+                if "do_not_run" in yaml_file.keys():
+                    external_vars.do_not_run = yaml_file["do_not_run"]
+                    if external_vars.do_not_run is None:
+                        external_vars.do_not_run = []
+                else:
+                    external_vars.do_not_run = []
+            else:
+                external_vars.known_failures = []
+                external_vars.failing_as_unspecified = []
+                external_vars.unsupported = []
+                external_vars.do_not_run = []
```
And another simplification. You can just use `in` on the dict, doesn't need to look at the `keys()` explicitly:

```
>>> d = {}
>>> 'foo' in d
False
>>> d['foo'] = 'bar'
>>> 'foo' in d
True
```

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,665 @@
+                external_vars.do_not_run = []
+
+
+def assign_environment_variables(classpath, vm, test_exceptions, generate_report):
+    """
+    Assign the environment variables needed for execution based on the name
+    classpath: String being the name of the variable CLASSPATH
+    vm: String being the name of the variable VM
+    test_exceptions: String being the name of the variable TEST_EXCEPTIONS
+    generate_report: String being the name of the variable GENERATE_REPORT
+    """
+    # Work out settings for the application (They are labelled REQUIRED or OPTIONAL)
+    if classpath not in os.environ:  # REQUIRED
+        sys.exit("Please set the CLASSPATH environment variable")
+
+    if vm not in os.environ:  # REQUIRED
+        sys.exit("Please set the VM environment variable")
+
+    if test_exceptions in os.environ:  # OPTIONAL
+        external_vars.TEST_EXCEPTIONS = os.environ[test_exceptions]
+
+    if generate_report in os.environ:  # OPTIONAL
+        # Value is the location
+        # Its prescense in env variables signifies intent to save
+        external_vars.GENERATE_REPORT = os.environ[generate_report]
+
+    external_vars.CLASSPATH = os.environ[classpath]
+    external_vars.VM = os.environ[vm]
```
Hm, I am not sure I see the benefit of having this a function.
Also, passing in the constants on call seems to make this just harder to figure out.

rhys-h-walker replied:
My initial idea was to convert this to a method call to be able to test the logic, however I've not thought of a good way of doing such so it will probably revert back to the old implementation.

smarr commented:
Hm, there's some problem with the current output.

Looks like everything is turned lower case:

```
E     stdout diff with stdout expected
E
E       0
E     - 1
E     - true
E     - false
E     - false
E     - false
E     - true
E     - false
E     - false
E     - false
E     stderr diff with stderr expected
E
E     + warning: a terminally deprecated method in sun.misc.unsafe has been called
E     + warning: sun.misc.unsafe::objectfieldoffset has been called by com.oracle.truffle.api.nodes.nodeclassimpl$nodefielddata (file:/users/smarr/projects/som/trufflesom/../graal/truffle/mxbuild/dists/truffle-api.jar)
E     + warning: please consider reporting this to the maintainers of class com.oracle.truffle.api.nodes.nodeclassimpl$nodefielddata
E     + warning: sun.misc.unsafe::objectfieldoffset will be removed in a future release
E     + exception in thread "main" org.graalvm.polyglot.polyglotexception: com.oracle.truffle.api.dsl.unsupportedspecializationexception: unexpected values provided for booleaninlinedliteralnode.andinlinedliteralnode@51549490: [1], [long]
E     + 	at trufflesom/trufflesom.interpreter.nodes.specialized.booleaninlinedliteralnode.evaluateargument(booleaninlinedliteralnode.java:48)
E     + 	at trufflesom/trufflesom.interpreter.nodes.specialized.booleaninlinedliteralnode$andinlinedliteralnode.executeboolean(booleaninlinedliteralnode.java:69)
E     + 	at trufflesom/trufflesom.interpreter.nodes.specialized.booleaninlinedliteralnode$andinlinedliteralnode.executegeneric(booleaninlinedliteralnode.java:63)
E     + 	at trufflesom/trufflesom.interpreter.nodes.sequencenode.executegeneric(sequencenode.java:48)
E     + 	at trufflesom/trufflesom.interpreter.method.execute(method.java:62)
E     + 	at <som> bool4>>#run(unknown)
E     + 	at <som> system>>#initialize:(smalltalk/system.som:48:1961-1963)
E     + 	at <som> (unknown)
E     + 	at org.graalvm.polyglot/org.graalvm.polyglot.context.eval(context.java:418)
E     + 	at trufflesom/trufflesom.launcher.eval(launcher.java:32)
E     + 	at trufflesom/trufflesom.launcher.main(launcher.java:13)
E     + original internal error:
E     + com.oracle.truffle.api.dsl.unsupportedspecializationexception: unexpected values provided for booleaninlinedliteralnode.andinlinedliteralnode@51549490: [1], [long]
```

rhys-h-walker commented:
> Hm, there's some problem with the current output.
> 
> Looks like everything is turned lower case:
> 
> ```
> E     stdout diff with stdout expected
> E
> E       0
> E     - 1
> E     - true
> E     - false
> E     - false
> E     - false
> E     - true
> E     - false
> E     - false
> E     - false
> E     stderr diff with stderr expected
> E
> E     + warning: a terminally deprecated method in sun.misc.unsafe has been called
> E     + warning: sun.misc.unsafe::objectfieldoffset has been called by com.oracle.truffle.api.nodes.nodeclassimpl$nodefielddata (file:/users/smarr/projects/som/trufflesom/../graal/truffle/mxbuild/dists/truffle-api.jar)
> E     + warning: please consider reporting this to the maintainers of class com.oracle.truffle.api.nodes.nodeclassimpl$nodefielddata
> E     + warning: sun.misc.unsafe::objectfieldoffset will be removed in a future release
> E     + exception in thread "main" org.graalvm.polyglot.polyglotexception: com.oracle.truffle.api.dsl.unsupportedspecializationexception: unexpected values provided for booleaninlinedliteralnode.andinlinedliteralnode@51549490: [1], [long]
> E     + 	at trufflesom/trufflesom.interpreter.nodes.specialized.booleaninlinedliteralnode.evaluateargument(booleaninlinedliteralnode.java:48)
> E     + 	at trufflesom/trufflesom.interpreter.nodes.specialized.booleaninlinedliteralnode$andinlinedliteralnode.executeboolean(booleaninlinedliteralnode.java:69)
> E     + 	at trufflesom/trufflesom.interpreter.nodes.specialized.booleaninlinedliteralnode$andinlinedliteralnode.executegeneric(booleaninlinedliteralnode.java:63)
> E     + 	at trufflesom/trufflesom.interpreter.nodes.sequencenode.executegeneric(sequencenode.java:48)
> E     + 	at trufflesom/trufflesom.interpreter.method.execute(method.java:62)
> E     + 	at <som> bool4>>#run(unknown)
> E     + 	at <som> system>>#initialize:(smalltalk/system.som:48:1961-1963)
> E     + 	at <som> (unknown)
> E     + 	at org.graalvm.polyglot/org.graalvm.polyglot.context.eval(context.java:418)
> E     + 	at trufflesom/trufflesom.launcher.eval(launcher.java:32)
> E     + 	at trufflesom/trufflesom.launcher.main(launcher.java:13)
> E     + original internal error:
> E     + com.oracle.truffle.api.dsl.unsupportedspecializationexception: unexpected values provided for booleaninlinedliteralnode.andinlinedliteralnode@51549490: [1], [long]
> ```

Output will be converted to lower case if case_sensitive is not set to be true in the test file. If case sensitivity is imperative for a test then it needs to be specified in the test file. The diff just reflects that, unless I am missing something in the output.

smarr commented:
> Output will be converted to lower case if case_sensitive is not set to be true in the test file. If case sensitivity is imperative for a test then it needs to be specified in the test file. The diff just reflects that, unless I am missing something in the output.

Hm, right. True, it's the diff.
I found it confusing seeing it like this. Hmmm...

rhys-h-walker committed:
Added ability to specify *** to require part of word on left and check if there for word on right.

smarr committed:
Fix PyLint issues and apply Black formatting

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Simplify code

- simplify assertions, pytest already shows that info
- make failures pytest.fail, and move reading of envvars into prepare_tests()

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@rhys-h-walker I squashed lots of commits, to have a better overview. I also rebased on the new master, that contains the merge with integration tests.

Please have a look at the commits I added, there were various bits that I pointed out earlier, and that simplify the code a bit.

Please also have a look at the test_tester stuff.
At the moment it depends on the directory pytest is executed in whether the tests pass or not.

This passes: `core-lib/IntegrationTests$ pytest -s -v test_tester.py -m tester`

But from one level up, it doesn't work: `core-lib$ pytest -s -v IntegrationTests/test_tester.py -m tester`

```
>       assert external_vars.known_failures == test_list
E       AssertionError: assert ['Integration...hod/test.som'] == ['./Tests/mut...hod/test.som']
E         
E         At index 0 diff: 'IntegrationTests/Tests/mutate_superclass_method/test.som' != './Tests/mutate_superclass_method/test.som'
E         
E         Full diff:
E           [
E         -     './Tests/mutate_superclass_method/test.som',
E         ?      ^...
```

rhys-h-walker commented:
> @rhys-h-walker I squashed lots of commits, to have a better overview. I also rebased on the new master, that contains the merge with integration tests.
> 
> Please have a look at the commits I added, there were various bits that I pointed out earlier, and that simplify the code a bit.
> 
> Please also have a look at the test_tester stuff. At the moment it depends on the directory pytest is executed in whether the tests pass or not.
> 
> This passes: `core-lib/IntegrationTests$ pytest -s -v test_tester.py -m tester`
> 
> But from one level up, it doesn't work: `core-lib$ pytest -s -v IntegrationTests/test_tester.py -m tester`
> 
> ```
> >       assert external_vars.known_failures == test_list
> E       AssertionError: assert ['Integration...hod/test.som'] == ['./Tests/mut...hod/test.som']
> E         
> E         At index 0 diff: 'IntegrationTests/Tests/mutate_superclass_method/test.som' != './Tests/mutate_superclass_method/test.som'
> E         
> E         Full diff:
> E           [
> E         -     './Tests/mutate_superclass_method/test.som',
> E         ?      ^...
> ```

Yep, that's an easy enough fix I think one of the testers is missing a relative path.

I have also noticed that the GENERATE_REPORT function no longer works correctly. The names it outputs for test include the whole core-lib/IntegrationTests so it no longer works directly as a TEST_EXCEPTIONS file. I'll fix that too

rhys-h-walker committed:
Further improvements

- support relative paths from CWD for testing
- GENERATE_REPORT is now compatible with multiple directories. It will remove anything in the path from before Tests/
- test exceptions no longer alters path to yaml file
- can now specify @tag_name in custom_classpath to load an environment variable by that name. Useful for custom classpath loading
- updated pytest parameterize to not run prepare_tests twice and IDs now match what is expected
- updated GENERATE_REPORT to have all tests with a consistent name from Tests/
- remove redundant assert messages
- split parse_test_file into methods, allows for more fine grained testing
- updated README
- added a new test for check_partial_word

rhys-h-walker commented:
Also, Tests/vector_awfy_capacity.som may need updating, it features a fixed classpath that is passed to the SOM++ interpreter.

```
VM:
    status: success
    custom_classpath: ./core-lib/Examples/AreWeFastYet/Core:./core-lib/Smalltalk
    stdout:
        50
        100
        10
```

It'll need some kind of  relative path location added to it

smarr commented:
> It'll need some kind of relative path location added to it

hm, yeah, I suppose that's a little different to the load_file case where it's inside the SOM program. Not sure whether we can find a consistent way of deal with both.

rhys-h-walker commented:
> > It'll need some kind of relative path location added to it
> 
> hm, yeah, I suppose that's a little different to the load_file case where it's inside the SOM program. Not sure whether we can find a consistent way of deal with both.

This issue is resolved, specify through @tag an environment variable which can be used in place for a custom_classpath.

```

AWFY=./core-lib/Examples/AreWeFastYet/Core
CLASSPATH=./Smalltalk

"
VM:
    status: success
    custom_classpath: @AWFY:@CLASSPATH
    stdout:
        nil

"

custom_classpath= ./core-lib/Examples/AreWeFastYet/Core:./Smalltalk
```

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,484 @@
+                            for tag in split_list:
+                                if "@" in tag:
+                                    tag = tag.replace("@", "")
+                                    classpath_joined += os.environ[tag] + ":"
```
Hm, this is unsafe code. It will completely break pytest when the environment variable isn't set, instead of just failing the test.

rhys-h-walker replied:
Yep it was, has been resolved in this commit:

[`7a024e7` (#128)](https://github.com/SOM-st/SOM/pull/128/commits/7a024e72e93a3f6ec033c8a710323f58b59c3289)

smarr commented on IntegrationTests/test_runner.py:
```diff
@@ -0,0 +1,484 @@
+                                classpath_joined += tag + ":"
+                        else:
+                            classpath_t = classpath_t.replace("@", "")
+                            classpath_joined += os.environ[classpath_t]
```
Same here. See for instance https://github.com/rhys-h-walker/SOMpp/actions/runs/16354125584/job/46208103214#step:9:75

smarr committed:
Revise code for maintainability

- turn test into parameterized ones, which gives errors that are more explicit about what’s going wrong
- rename and restructure various bits
- rewrite parser of test definitions to be a more classic single-pass parser
- reify test definition as object, and make env var failures test definitions that fail as a test
- avoid opening and reading file content when detecting tests. First, we find candidates, and then we check whether we can parse them.
- build the error message from the test definition
- test and handle case sensitive setting
- use an object for report details, instead of a bunch of globals on the module
- rename envvar to TEST_EXPECTATIONS
- use $ to identify envvars in the test configuration
- rename envvar to GENERATE_EXPECTATIONS_FILE
- revise README.md

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[GHA] Allow setting a specific test expectations file

- SOM-RS shouldn’t need custom test command anymore
- JAVA_HOME for TruffleSOM should already be set by the build step
- set env vars for PySOM in build step
- use GITHUB_ENV file for relevant variables

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Change load_file.som to use a file in TestSuite, which is usually symlinked

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
[GHA] Mark SOM-RS as not up to date

It’s a little buggy…

Signed-off-by: Stefan Marr <git@stefan-marr.de>
